### ISSUE TITLE
API shouldn't require a username and password

### ISSUE THREAD

**Issue by anarcat** [wallabag/wallabag#2800]
### Issue details

The API currently requires new "API client" to be created to operate, which includes a `client_id` and a `client_secret`.

But to *actually* work, the API *also* needs an `access_token` and `refresh_token`. And *those* are very short lived - one hour maximum.

This means to implement a client, you need to not only ask the user for an incomprehensible `client_id` and `client_secret`, but *also* for their username and password. Every client then needs to *store* those username and passwords indefinitely to regenerates the access tokens.

### Environment

* wallabag version: 2.1.0 API documentation
* How did you install wallabag? irrelevant.
* Last wallabag version that did not exhibit the issue (if applicable): N/A
* php version: N/A
* OS: N/A
* type of hosting (shared or dedicated): N/A
* which storage system you choose at install (SQLite, MySQL/MariaDB or PostgreSQL): N/A

### Steps to reproduce/test case

N/A.

### Suggested solution

The username and password should be used only to create access tokens, which should be permanent. This would make the whole API way more efficient because then a single HTTP request would be require for API calls instead of two (assuming the worst case of less than one request per hour per client).

Alternatively, client tokens should  be user-specific. I believe there may already be an issue about this, but right now, anyone can see (*and* delete?) anyone else's tokens on the server. That seems like a Real Bad Idea.

Clients should only ask the user's username and password as a transient secret, that can be used to create client tokens on the fly, which then can be used to create access tokens.

**Comment by j0k3r** [comment]
> Alternatively, client tokens should be user-specific

It's already in place in the 2.2.0 (see https://github.com/wallabag/wallabag/pull/2499)

And we need to discuss about your interesting point.

**Comment by anarcat** [comment]
from what i see in #2806 - token ownership is not actually checked?

so there would still be work needed there.

**Comment by jlnostr** [comment]
I agree. Of course, you can refresh the tokens after they expired but in the case something fails after submitting the request, the new tokens are lost forever and the user needs to be asked for their username and password again. The suggested solution by @anarcat seems good to me.

Otherwise issues like https://github.com/wallabag/windows-app/issues/32 can happen at any time.

**Comment by jlnostr** [comment]
BTW, I have a question, mainly to the core team. My windows app is using the username and the password once to generate the `access_token` and the `refresh_token`. Those both values are saved in a database. However, I got some users reporting me that the app went from useful to useless after some weeks. I found out that this happens due to a failing refresh request, where my app might not get the newest values, even if the refresh request was done before. To illustrate this issue: 

```
tokens are outdated -> Send request to server -> lose internet connection -> Server response is not getting parsed, but the old tokens are nevertheless invalid
```

The only solution would be to save the username and the password as well and use them to generate a fresh pair of tokens in case the above schema failed. Is that okay for you?

**Comment by tcitworld** [comment]
Nope, that's a no-go in terms of security. Let's increase the refresh token lifetime instead.

**Comment by anarcat** [comment]
how about making the refresh token permanent and revokable?

**Comment by tcitworld** [comment]
Not permanent IMHO. If the user didn't log inside the app for something like a year, you can expect the app to ask the user again for credentials.

**Comment by anarcat** [comment]
my app can't ask for credentials. it's automated.

most apps require username and password only when setting up, not after an arbitrary timeout. as far as i know, twitter Oauth credentials are permanent, for example.

the title of the issue i opened here is "API shouldn't require a username and password", and I meant it. it's not "API shouldn't require a username and password most of the time". i want to have a token i can use until revoked.

why are you expiring the tokens anyways? what's the use case for this?

if we are worried that users will lose that device, then the proper solution is *not* to expire the tokens, because then we are vulnerable for a long time after the device is lost: if i lose the device N days after then tokens are used, then I am vulnerable for full year.  the proper solution is, as done elsewhere, to have a clear interface to see what devices have access and allow revocation.

**Comment by jlnostr** [comment]
I agree with @anarcat. This is the best way for the future and should be included in wallabag 2.3. Sooner is better. What should we do for older versions? Should we log out the user and enforcing him to re-login again?

**Comment by di72nn** [comment]
The problem arises from the fact that the only allowed `grant_type` (`grant_type=password`) configured to require a weird set of parameters: it takes both username and password (ok for this grant type) **and** `client_id` and `client_secret`.
**`client_id` and `client_secret` are supposed to be used with `grant_type=client_credentials` which doesn't need username and password.**

**If `grant_type=client_credentials` is allowed, all the other problems discussed here are gone.**

See: [OAuth 2 Simplified](https://aaronparecki.com/oauth-2-simplified/) (or read other OAuth-related info).

**Comment by di72nn** [comment]
There is `authorization_code` `grant_type` which should also be a good choice in case of wallabag, but:
 1. There is no endpoint to get that authorization code.
 2. It has pretty much the same disadvantages discussed here (it allows to get tokes **once**; if the tokens expire, the procedure needs to be repeated).

**Comment by anarcat** [comment]
from my point of view, we should be able to get as many tokens as we want with the `client_id` and `client_secret`. it seems completely redundant to have to provide the username and password as well, *unless it is to **create** a new `client_id` and `client_secret` pair*!

in other words, maybe the simplest solution for clients would be to have a user-specific `client_id` and `client_secret` that can be used (without passwords!) to generate temporary tokens on the fly. then the `client_id` and `client_secret` tokens could be used to revoke access... 

obviously, for this to work, older pre-upgrade tokens would need to be purged or categorized differently, because they are currently publicly readable among all users.

is that what you meant?

**Comment by di72nn** [comment]
This is how it currently works:
```
http POST http://v2.wallabag.org/oauth/v2/token \
    grant_type=password \
    client_id=12_5um6nz50ceg4088c0840wwc0kgg44g00kk84og044ggkscso0k \
    client_secret=3qd12zpeaxes8cwg8c0404g888co4wo8kc4gcw0occww8cgw4k \
    username=yourUsername \
    password=yourPassw0rd
```

This is how it should work:
```
http POST http://v2.wallabag.org/oauth/v2/token \
    grant_type=password \
    client_id=12_5um6nz50ceg4088c0840wwc0kgg44g00kk84og044ggkscso0k \
    username=yourUsername \
    password=yourPassw0rd
```
Notice the absence of `client_secret` - `grant_type=password` shouldn't require it. Sources: [simplified](https://aaronparecki.com/oauth-2-simplified/#password), [RFC](https://tools.ietf.org/html/rfc6749#section-4.3).

In theory API clients (read: `client_id` and `client_secret` pairs) should not be user-specific, but since they already are in recent versions, this is what may be added:
```
http POST http://v2.wallabag.org/oauth/v2/token \
    grant_type=client_credentials \
    client_id=12_5um6nz50ceg4088c0840wwc0kgg44g00kk84og044ggkscso0k \
    client_secret=3qd12zpeaxes8cwg8c0404g888co4wo8kc4gcw0occww8cgw4k
```
Sources: [simplified](https://aaronparecki.com/oauth-2-simplified/#client-credentials), [RFC](https://tools.ietf.org/html/rfc6749#section-4.4).

For f*ck's sake, people, read the OAuth specs before inventing more crap.

**Comment by jlnostr** [comment]
I've added this to the 2.3 milestone and dropped a bounty, cause it's a very important issue to me. Hopefully someone will find the time to fix it.

[![Bountysource](https://api.bountysource.com/badge/issue?issue_id=41448076)](https://www.bountysource.com/issues/41448076-api-shouldn-t-require-a-username-and-password?utm_source=41448076&utm_medium=shield&utm_campaign=ISSUE_BADGE)

**Comment by tcitworld** [comment]
Did add a bounty too since I'm not sure about having the time.

**Comment by Eisfreak7** [comment]
If this requires a significant revamp of the authentication handling maybe it would also be a good time to look into scopes.
An example use case would be a service like [ifttt](https://ifttt.com/). Such a service would need the permission to add new articles, but not permissions to read or modify my list of existing entries.
For reference: https://www.brandur.org/oauth-scope

**Comment by jlnostr** [comment]
@Eisfreak7 I think that's another story. Could you please create a new issue for that?

**Comment by j0k3r** [comment]
Looks like @tcitworld won the bounty :slightly_smiling_face: :moneybag:

**Comment by tcitworld** [comment]
I'll put it back on https://github.com/wallabag/wallabag/pull/3068

**Comment by tcitworld** [comment]
Put back the bounty on since this isn't fixed after all.

**Comment by tcitworld** [comment]
And an extra one since I'm to blame for this failure.
[![](https://api.bountysource.com/badge/issue?issue_id=41448076)](https://www.bountysource.com/issues/41448076-api-shouldn-t-require-a-username-and-password)

**Comment by alexandery** [comment]
I'm here because Firefox add-on for Wallabag requires 4 pieces of information, including my username and password for the instance of Wallabag. WHAT? The whole purpose of client_id and client_secret is NOT to require credentials. I DO NOT want to provide my credentials to any client app that wants to connect - that makes zero sense. My account is vulnerable if I keep leaving my credentials in all apps. And @anarcat has expressed exactly that.

So, what's the outcome of such a long discussion? It seems this is not going to change? Would like to better understand what you guys are planning for this.

**Comment by Rurik19** [comment]
The quickest solution may be to set **refresh token expiration time** as long as possible - instead of 2 weeks, let say, 10 years. In that case, an application can get app/refresh token pair only first time and then refresh app token (almost :smile: ) forever and not store login/password. I can make wallabagger extension version with no stored credentials in 5 minutes.

**Comment by deanishe** [comment]
Clients should never even *see* a user's login credentials, tbh. That's the whole point of OAuth, after all.

The client should redirect the user to the Wallabag server where they log in and grant the client permission (or not). The client then receives the tokens it needs from the server without ever seeing the user's password.

That's perhaps not practical for some kinds of clients that can't start a local webserver to receive the tokens, but that should be the default mode of operation.

Requiring users to give clients their passwords essentially defeats the whole purpose of using OAuth in the first place.

In that case, it would be better (simpler and more secure) to just let users generate (and revoke) an API key for each client that is used in place of username+password.

**Comment by vext01** [comment]
Just wanted to say: I'm evaluating wallabag as a pocket alternative and this security issue is the largest barrier to adoption for me right now.

**Comment by Strubbl** [comment]
This issue is really ugly to have

**Comment by anarcat** [comment]
quite frankly, if the Oauth implementation in Wallabag is going to stay this way, it might as well be removed, because it just makes writing clients needlessly hard at absolutely no security advantage.

**Comment by virtadpt** [comment]
For what it's worth, I had to farm it out to a bot because it was too cumbersome to do it any other way.

**Comment by alexandery** [comment]
@virtadpt Care to elaborate on your solution?

**Comment by virtadpt** [comment]
Source code here: https://github.com/virtadpt/exocortex-halo/blob/master/web_index_bot/send_url_to_wallabag.py

**Comment by tcitworld** [comment]
Hey folks. For what it's worth absolutely nobody is happy with the current situation, but let's keep productive in the comments so that we can fix this properly.
Of course everyone is welcomed to propose some pull request to fix the current authentication flow.

**Comment by deanishe** [comment]
> but let's keep productive in the comments so that we can fix this properly.

You already have two sensible, and much better, solutions suggested. What more do you need to start fixing it?

**Comment by anarcat** [comment]
> You already have two sensible, and much better, solutions suggested. What more do you need to start fixing it?

This is not useful. @tcitworld was pretty clear about what's needed: they need someone to do a PR and clean that code up. Snarky comments won't help anyone here... 

And I'm sorry if i seemed to be in the snarky crew. ;)

**Comment by deanishe** [comment]
> @tcitworld was pretty clear about what's needed

That he wants someone to do a PR. Which is fine. But why ask people not to post their workarounds in the thread when it's the only progress being made on the issue?

**Comment by anarcat** [comment]
> But why ask people not to post their workarounds in the thread when it's the only progress being made on the issue?

I don't think that's what @tcitworld was refering to at all.

**Comment by thibaultamartin** [comment]
Hej everyone, very interesting topic here. To sum-up, the ideal situation would be to have an authentication à la Nextcloud or Mastodon: when your client needs to log to any instance, you never need to type in a `client_id` or a `client_secret`, but you only need to type in your credentials.

# How OAuth2 works

Basically, OAuth2 needs to authenticate both the client application and the user to deliver an OAuth2 Token. The client application is authenticated thanks to the `client_id` and `client_secret` fields (as defined in the OAuth2 standard), and the user with its username and password (although OAuth2 does not define how the user is authenticated). A client application can use one of several flows to provide those applications to the _Authorization Server_ (_i.e._ the server part that issues tokens)

* Authorization Code + PKCE
* Implicit (deprecated)
* Resource Owner Password Credential (deprecated)
* Client Credential 

# How Wallabag uses OAuth2

The way Wallabag works at the moment, it uses the _client credentials_ flow. This means the user gives his crendentials to his client app. Then the client app sends all the credentials (user credentials _and_ `client_id`/`client_secret` couple). In exchange for this, the server answers with an _Access Token_, which is the one actually use to authenticate API calls; and a _Refresh Token_, used to renew Access Tokens when they expire, without needing to ask the user his usernam or password again.

# Two main issues

## Refresh Token lifetime

What seems to be an issue right now is that the Refresh Token is too short lived. Extending the Refresh Token life to a few months doesn't seem indecent to me: Wallabag is not a critical app and shouldn't be the main target in case of threat.

## client_id and client_secret configuration

The second most notable issue is that the user needs to create a `client_id` and `client_secret` in his administration interface, and provide it manually to the client application. This can be a real usability problem, especially with mobile clients.

The `client_id` and `client_secret` creation work can be delegated to the client application itself. This can be achieved thanks to the [RFC7591 - Dynamic Client Registration](https://tools.ietf.org/html/rfc7591).

I need to dig the specifications a bit more in depth, but in my opinion it is not useful to switch from the _Client Credentials_ flow to the _Authorization Code + PKCE_ flow. Indeed, the _Authorization Code + PKCE_ flow relies on a browser/web view to let the user input his username and password directly on the Authorization Server. The main reason for that is so the client application never sees the username and password, and so the user can delegate only a subset of his authorizations to his client application.

In the case of Wallabag, I don't see any issue if the open source clients see the username and password, and functionally speaking "delegating a subset of its authorization to the client app" does not make any sense.

**Comment by deanishe** [comment]
> I don't see any issue if the open source clients see the username and password

I do: it doesn't fix the core problem with the current implementation. Having to give someone your password is one of the main things tokens (including OAuth) are intended to avoid. It might not be a big deal giving the open-source iOS app your password, but what if you want to integrate your Wallabag with some other service?

The current OAuth implementation is very broken, but any replacement that still requires giving clients your password isn't a fix.

OAuth itself is fundamentally a poor solution for Wallabag because it's just not intended for this kind of application. The advantages of OAuth are:

1. Very simple for users. The application provider jumps through a few extra hoops (client IDs & secrets) so users only need to click a link and then "Allow" to authorise a client.
2. Allows applications to use third-party auth providers ("Log in with Facebook" etc.)
3. Client applications use tokens. Permissions can be granular and they never see users' passwords.

(2) is irrelevant here because Wallabag is the auth provider.

(1) becomes a disadvantage with Wallabag because its self-hosted nature means the user is also the provider. All complexity accrues to the user, which the current implementation demonstrates very well.

(3) is the only real advantage OAuth would have for Wallabag, and it's defeated when the user has to give a client their password.

And (3) can be achieved more simply (in terms of both implementation and user workflow) with simple API keys.

The user generates an API key in their Wallabag account and then gives it (and the URL of their Wallabag instance) to the client application. It's slightly more complex than giving the client your username and password, but far more secure.

In the context of Wallabag, it has all the advantages of OAuth without the disadvantages/complexity. The API keys are tokens in exactly the same way as OAuth tokens are, with all the same advantages, but without the complexity of OAuth (which has no additional benefit for Wallabag).

**Comment by Rurik19** [comment]
As I said a couple of times already - there is simple solution to problem of storing user credentials locally -- if refresh token will lasts forever then login | password requires only in the first time and no need to store them.

**Comment by deanishe** [comment]
> if refresh token will lasts forever then login | password requires only in the first time and no need to store them


But you still have to give them your password (and trust that they don't hold onto it). And it's entirely unnecessary. This is a solved problem. There is no reason you should have to give any client your password.

**Comment by Rurik19** [comment]
Why my password. Client use his own password, and in my suggestion  - only one time.

**Comment by deanishe** [comment]
> Why my password? 

Because they need your password to log into your account.

If you give each client its own password, that's just using API keys but with extra steps.

**Comment by Rurik19** [comment]
so wallabag should works - each client registers separately and use its own credentials to access it. Nobody need admins password.

**Comment by deanishe** [comment]
> each client registers separately and uses its own credentials to access its owns accounts.

How does a client read my bookmarks if it's not accessing my account?

**Comment by thibaultamartin** [comment]
Hej @deanishe, you have fair points here.

> but what if you want to integrate your Wallabag with some other service?

This is fair, and the whole point of OAuth2 is to be able to delegate a subset of authorizations to a client app without the client app knowing your username and password (which would give it access your whole account de facto).

Getting back to all your points:

(1) Can be entirely transparent for the user by implementing the very OAuth2 RFC [Dynamic Client Registration](https://tools.ietf.org/html/rfc7591). This means the user does not need to handle a `client_id` and `client_secret` at all. It only needs to type in its instance url, its username and its password.

(2) One of the other advantages of OAuth2 is that the identity provider not necessarily is Facebook or Google. It can be your organization's identity provider (e.g. Keycloak). That would allow SSO for organization which need it. Although, to be fair Wallabag doesn't seem to be your average corporate solution. It would be useful mostly for non-profits.

(3) Yes, an Access Token allows you to do grant a subset of your authorizations to a client app. An API Key could allow you to do the same, but it would require the end user to generate the API Key (most likely a long string, such as a client_id or a client_secret), and use it in its client app. If we use OAuth2 with dynamic client registration there is no long string to paste to the client app. Also, the Access Token has a limited lifetime, and can be refreshed with a Refresh Token. If the Refresh Token expired too, it still is possible for the user to simply log in again with its username and password. API Key rotation is a bit more cumbersome for the end user.

So I definitely think the best option is to implement both the _Authorization Code + PKCE_ flow (so the client app can be unaware of the credentials) and the _Client Credentials_ flow (so the legacy app can still easily authenticate the user). Those solution do not require the user to do any tweaking or API Key generation and pasting. They only require the user to know his very own credentials. This solution is used by Mastodon and Nextcloud, and work kinda well!

I agree the ideal solution would be to use _Authorization Code + PKCE_ only, as is currently recommended by the IETF. But I also think a pragmatic approach that takes into consideration the legacy clients is useful. Let's not forget that most people work on their free time on Wallabag and this issue is stalled since 2017 :)

**Comment by Rurik19** [comment]
> How does a client read my bookmarks
if you give them your credentials - it is like one user works from many places. All saves in one account (yours)

**Comment by deanishe** [comment]
> It only needs to type in its instance url, its username and its password

Which is the problem, imo. The `client_id` and `client_secret` are a bit annoying. But it's having to give a client application my password that is the actual problem. If Wallabag did the normal OAuth thing and sent users to the Wallabag installation to log in and grant the client access, that would be fine, too.

> That would allow SSO for organization which need it.

For SSO, Wallabag would first need to allow users to register via OAuth.

> API Key rotation is a bit more cumbersome for the end user.

There's no need to rotate API keys. The API key is either the refresh token or both tokens. The point of having an access and a refresh token is so that clients don't have to hit the authentication servers on every request. That's a big deal when the auth server is a separate or third-party service, but Wallabag does all of that itself.

**Comment by deanishe** [comment]
> if you give them your credentials

You literally just said that they don't get my credentials. The client gets its own instead.

Can you explain in detail what you mean? As far as I can tell, you just said the opposite of what you previously said.

**Comment by thibaultamartin** [comment]
> Which is the problem, imo. The client_id and client_secret are a bit annoying. But it's having to give a client application my password that is the actual problem. If Wallabag did the normal OAuth thing and sent users to the Wallabag installation to log in and grant the client access, that would be fine, too.

Then we agree. Dynamic client registration allows to skip the `client_id` and `client_secret` part, and _Authorization Code + PKCE_ is the "redirect the user to wallabag so the client app does not need to have the user's password".

> There's no need to rotate API keys. The API key is either the refresh token or both tokens. The point of having an access and a refresh token is so that clients don't have to hit the authentication servers on every request. That's a big deal when the auth server is a separate or third-party service, but Wallabag does all of that itself.

Well yes and no. From a security standpoint, API Key rotation could be something desirable, since an API Key can be leaked. Also, if you want to give granular access to some of your authorizations, the `scope` claim in OAuth2 Access Tokens has everything you need. If you want to go the API Key way, you need to implement the logic that binds an API Key to some authorizations. I personally find this a bit less elegant.

Functionnally speaking and from an end-user perspective, I definitely think that OAuth2 with dynamic client registration _and_ the _Authorization Code + PKCE_ is preferable.

**Comment by thibaultamartin** [comment]
To add a bit of documentation on the _Authorization Code_ flow, [auth0 docs](https://auth0.com/docs/flows/concepts/auth-code) are quite good.

Just note that in Wallabag context the "Auth0 Tenant" would be Wallabag itself.

**Comment by Rurik19** [comment]
> Can you explain in detail what you mean? 

1) you are the admin of Wallabag. you set up application, you have yours credentials = loginA / passA
2) there are users U1 .... Un. they have it's own's accounts and credentials = ( u1 / passU1)
3) U1 clients application needs to have u1 / passU1 for work
4)  clients doesn't need to have admin's **loginA / passA**, if they have it, that's admin's fault

**Comment by deanishe** [comment]
> Then we agree.

Not entirely. I think that solution is acceptable. I still prefer plain API keys because they are much easier to implement (vs OAuth client) and much better suited to CLI applications.

> API Key rotation could be something desirable, since an API Key can be leaked.

True. But it's hardly a great problem to regenerate an API key in such rare situations.

> the scope claim in OAuth2 Access Tokens has everything you need.

It's a small part of what you need. Associating a few permissions with a key is simple. 95% of the work is managing the permissions and implementing the authorisation, which OAuth does not help with.

> Functionnally speaking and from an end-user perspective […]

Functionally, there's little difference. I agree it's better for connecting some kinds of applications, but it's rather more complex for others. It also makes implementing clients more complex because OAuth is complex.

**Comment by deanishe** [comment]
> U1 clients application needs to have u1 / passU1 for work.

And that's exactly the problem that tokens are supposed to solve. Nobody but you was talking about admin credentials, as far as I can tell.

And of course, if you wanted to implement an admin API, then you *would* need to give clients the admin's credentials under your model.

**Comment by thibaultamartin** [comment]
Hej everyone, so a couple of weeks after stopping the discussion, I'm back with the best practices regarding OAuth2 and what should be done.

# Types of clients

The first step is to know which are the types of clients that will be using OAuth2 Access Tokens. The short answer is: all of them. The longer answer is:
* The web-based interface of the instance itself. While it does not use OAuth2 today (or does it?), it would be a nice to have. It would be useful to simplify the process of companion apps. More on that below
* The native mobile apps ([iOS](https://github.com/wallabag/ios-app), [Android](https://github.com/wallabag/android-app)) or [GTK Desktop and Mobile app](https://gitlab.gnome.org/bilelmoussaoui/read-it-later)
* A potential evolution of the frontend of the Wallabag instance frontend, if it ever were to become a single page application (SPA)

**The IETF's [best practices for Native Apps](https://tools.ietf.org/html/rfc8252#section-4) and [best practices for Browser Based Apps](https://tools.ietf.org/html/draft-ietf-oauth-browser-based-apps-04#page-9) recommend using the _Authorization Code + PKCE_ flow for these client apps.**

# Registration of the client app

In order for any client app to be able to use OAuth2 flows on Wallabag, it needs to be registered on the instance. Today, that registration is manually handled by the end-user, ho creates new client apps. Wallabag generates a `client_id` (which is basically the client app's username) and a `client_secret` (which used to be the client app's password, but is kinda legacy now).

What we want now is the ability to use dynamic client registration. This will allow the client app to reach the instance, and ask it to generate a `client_id` and a `client_secret`. During the process, the client app tells the server to bind its `client_id` to a `redirect_URI`. The instance does not do any check with that URI, it simply stores it and binds it to the `client_id`. More on that redirect URI below.

# Authorization Code + PKCE

## Authorization Request

Let's put aside the PKCE bit for the sake of clarity. The Authorization Code flow makes use of the end-user's browser to handle the actual end-user authentication. The following describes how to handle this authorization flow for a native app (most complicated case). We'll discuss other cases later.

![image](https://user-images.githubusercontent.com/1688309/75119033-a65a6a00-567f-11ea-9169-04ee1362ed6a.png)

Mermaid source for this diagram:

```
sequenceDiagram
    Wallabapp->>+Browser: Open user browser on Wallabag server's authentication page
    Browser->>Wallaserver: Get login page
    Wallaserver->>Browser: Send login page
    Browser->>Wallaserver: Send username and password
    Wallaserver->>Browser: Send Authorization Code
    Browser->>-Wallabapp: Retrieve Authorization Code
    Wallabapp->>Wallaserver: Exchange Auth Code against tokens
    Wallaserver->>Wallabapp: Send Access Token and Refresh Token
```

This clever use of the user's browser allows the client app to retrieve an Access Token without ever being aware of the end-user's credentials. We'll discuss the use of the Access and Refresh Token below. We can see that the end-user's browser seems to call the native app. While this may seem odd, this can be performed thanks to the _redirect_URI_.

A _redirect_URI_ is a URI bound to the client app. Wallabag client apps should ask to use `wb://redirect` as a redirect_URI. When the Wallabag instance uses an http redirection in the end-user's browser to the _redirect_URI_, it uses query parameters. As an example, if the instance redirects the browser to `wb://redirect?lorem=ipsum`, the end-user's browser will parse the url, open the app associated with the `wb://` scheme and pass it the URL. Our client app will be associated with the `wb://` scheme, get the URL, and be able to determine the value of `lorem` to be equal to `ipsum`.

Each client app is bound to the redirect_URI it specified during its registration against the server.

Now let's get a bit more technical with the following diagram.

![image](https://user-images.githubusercontent.com/1688309/75119855-fd177200-5686-11ea-8307-bdfe4c6fa2f2.png)

```
sequenceDiagram
    Wallabapp->>+Browser: Open user browser on Wallabag server's authentication page
    Browser->>Wallaserver: (1) GET /authorize
    Wallaserver->>Browser: Send login page
    Browser->>Wallaserver: (2) POST /authorize
    Wallaserver->>Browser: (3) Send Authorization Code to redirect URI
    Browser->>-Wallabapp: Retrieve Authorization Code
    Wallabapp->>Wallaserver: (4) POST /token
    Wallaserver->>Wallabapp: (5) Send Access Token and Refresh Token
```

The app displays a field for the end-user to type in the instance URL. Once the end-user has clicked/tapped the button, the flow (1) starts: the app launches the user's browser _but not an web widget/web view_ with the following request

`GET /authorize?response_type=code&client_id=`_client_id_`&state=`_state_parameter_

For the sake of simplicity, let's consider that the _state_parameter_ is a random string generated and stored by the client app, which will be used later. See [Auth0's article on that matter](https://auth0.com/docs/protocols/oauth2/mitigate-csrf-attacks). The actual best practices are a bit different. See also [the IETF specification for that flow (1), called the _Authorization Request_](https://tools.ietf.org/html/rfc6749#section-4.1.1).

The Wallabag instance answers with an authentication page, displaying two fields: one for the user login, and one for the user password, plus a validation button. Once the end-user has filled the fields, he sends the form, which sends the following request:

```
POST /authorize`wb://`_redirect_URI_`

username=endusername&password=enduserpassword
```

## Authorization Response

The Wallabag instance checks the user credentials in its internal database, and if the credentials are valid, it generates an _authorization code_ and redirects the end-user's browser to the `redirect_URI` bound to the client app that initiated the Authorization Request. It adds the `code` and `state` query parameters to that URI. As an example, the response could be a redirect to `wb://redirect?code=1bsf4gqa&state=a2fs7d35`

See [the IETF specification for the flow (3), called the _Authorization Response_](https://tools.ietf.org/html/rfc6749#section-4.1.2)

## Access Token Request

The client app retrieves the url from the redirect of the Authorization Response by parsing the query parameters. It checks that `state` is consistent with the `state` sent in the Authorization Request, and retrieves the `code` parameter.

It then calls the `/token` endpoint of the instance with the following query parameters

* `grant_type` which must be equal to `authorization_code"
* `code` which must be equal to the `code` parameter retrieved from the Authorization Reponse
* `redirect_uri` which must be the same redirect URI as used when registering on the server
* `client_id` which must be its own client_id

See [the IETF specification for this flow (4) called the Access Token Request](https://tools.ietf.org/html/rfc6749#section-4.1.3)

## Access Token Response

The Wallabag instance checks all the parameters, and if successful replies with an Access Token and a Refresh Token, along with the Access Token's time to live.

See [the IETF specification for this flow (5) called the Access Token Response](https://tools.ietf.org/html/rfc6749#section-4.1.4).

---

I'll get a bit more in details in the dynamic registration of the client app in another comment. I can assist anyone willing to implement the Authorization Code flow, but my knowledge of PHP and Symfony is too low at the moment to produce actually clean and maintainable code.

**Comment by deanishe** [comment]
> The web-based interface of the instance itself. While it does not use OAuth2 today (or does it?), it would be a nice to have.

An app can't be its own OAuth client. It's a mechanism for logged-in users to grant access credentials to *3rd-party* client applications. Unless you also want to make Wallabag an OAuth *client* to other providers ("Log in with Google", "Log in with Facebook" etc.), it needs its own login mechanism separate from OAuth.

I think there is some confusion here between OAuth clients and providers. And overall a bit too much focus on the OAuth workflow without considering how that fits into Wallabag's needs.

**Comment by thibaultamartin** [comment]
> An app can't be its own OAuth client. It's a mechanism for logged-in users to grant access credentials to 3rd-party client applications.

In the classical model, the Authorization Server and the Resource Server are two separate entities. Wallabag being self-hosted, it does not make sense to ship a separate Authorization Server. Merging the Authorization Server in Wallabag backend does make sense :)

GMail is an example of such a behaviour: when going unlogged to gmail.com, you are redirected to accounts.gmail.com (the AS of Google) so you can authenticate.

I can't see a reason off the top of my head why we should maintain two separate authentication mechanisms, and how Wallabag using its built-in AS for the authentication on the web interface would be a problem? Especially if in the long run that web frontend is a SPA

To make thinks a bit clearer:
* Wallabag would ship with a built-in AS (my previous message is about which endpoints should be implemented to provide such an AS)
* That built-in AS needs to provide an authentication form (with login and password fields, although the OAuth2 specification does not specify how the authentication should be handled). This page comes from the built-in AS of Wallabag, and is presented to any user who is not logged in. The username and password are the Wallabag user's username and password. They are checked against the Wallabag database.

> I think there is some confusion here between OAuth clients and providers. 

I'm quite positive the roles of OAuth2 clients and providers are well defined here :)
Wallabag definitely would be a provider, since we don't want to rely on social login. The wallabag apps would be OAuth2 clients. The web frontend can be client too, especially if in the long run it becomes a SPA.

> And overall a bit too much focus on the OAuth workflow without considering how that fits into Wallabag's needs.

Well, basically we want:
* A way for a user to delegate access to his account to a 3rd party client without this 3rd party client to be aware of his password ▶️ OAuth2 seems to be the way to go
* A way for a user to log-in by typing his username and password, and nothing else ▶️ Dynamic client registration and Auth Code + PKCE seems to be the way to go

This topic is pretty interesting. If you have suggestions regarding how to handle it differently I'd gladly discuss that

**Comment by deanishe** [comment]
> I can't see a reason off the top of my head why we should maintain two separate authentication mechanisms

Which is why making Wallabag's own frontend an OAuth client is a bad idea. One way or another, Wallabag's own frontend *must* support the native login mechanism. So why add a second (OAuth) to it?

The rest of the API only cares about the access token. It doesn't matter whether the client got it via OAuth, via Wallabag's native login or the user explicitly generated it in their preferences. 

OAuth is only a concern for the client and the authentication-related parts of the API. Everything else only cares about the access token.

> although the OAuth2 specification does not specify how the authentication should be handled

That's because OAuth is only concerned with *third-party* authentication or authorisation. It has nothing to do with logging into the service itself.

> I'm quite positive the roles of OAuth2 clients and providers are well defined here :)

You were talking about OAuth helping with SSO earlier in the thread. That only makes sense in the context of Wallabag being the OAuth *client*, unless you think people might use Wallabag as their organisation's SSO server, which seems rather out-of-scope.

I'm not sure how to put this tactfully … You seem to be rather caught up in the details of OAuth without having a solid concept of how it fits into the big picture.

OAuth is only good for authorising 3rd-party applications and webapps. It's pointless for authorising your own web frontend (which must support the native mechanism) and very unpleasant for authorising command-line programs/libraries.

**Comment by thibaultamartin** [comment]
> One way or another, Wallabag's own frontend _must_ support the native login mechanism.

Why is that the front end must sport a native login mechanism?

In companies, you have an AS which is not merged to the RS the same way we would do it with Wallabag. The only “authentication front end” is the authentication page the AS presents to the end user. The user has a “company account” and he delegates access to parts of his account to different apps in the company. All those apps are OAuth2 clients.

In the specific context of Wallabag, which is a self-hosted application, we do not deploy a separate AS but would build it in the RS. This means the wallabag instance would be both an AS and an OAuth2 client.

> > although the OAuth2 specification does not specify how the authentication should be handled
> 
> That's because OAuth is only concerned with _third-party_ authentication or authorisation. It has nothing to do with logging into the service itself.

Well it simply means you can login on the AS with a username and password, a U2F key, a session cookie, Kerberos, whatever suits your need.

> You were talking about OAuth helping with SSO earlier in the thread. That only makes sense in the context of Wallabag being the OAuth _client_, unless you think people might use Wallabag as their organisation's SSO server, which seems rather out-of-scope.

To put it simply:
* You go to your instance URL
* It redirects you to the built-in AS which displays an authentication page
* You log in with your username and password
* This saves a session cookie in your browser
* You are redirected to the wallabag frontend with an Access Token which proves that you authenticated successfully
* You do some wallabaging
* You then decide to install Read It Later (GTK client app)
* You launch Read It Later which prompts you for your instance URL
* You type in your instance URL and validate
* This launches your browser to Wallabag built in AS’ authentication page
* You already have a cookie for that page so you don’t need to authenticate with your login and password (which is why the specification doesn’t mention how you must authenticate against the AS)
* I’m skipping the scopes/consent part here
* Your browser uses the redirect URI to call back Read It Later, which can then retrieve the Access Token

SSO may not be the exact right term here, but you effectively type in your username and password only once to log in to two clients (the Wallabag instance frontend and the Read It Later client).

I feel we’re getting into a purists argument around whether the Wallabag frontend should be considered as an OAuth2 client or not, which doesn’t really matter in the end.

If the frontend keeps its own authentication interface independant of OAuth2, but we implement the built in AS for third parties client, then you can:
* delegate third parties access to your account
* without said third parties to be aware of your password
* without requiring the user to know anything else than his username and password

If the frontend is considered as an OAuth2 client too, you can:
* do all of the above
* plus adding new third party clients on the same machine doesn’t require you to log in again

**Comment by elfrinjo** [comment]
I guess becoming a full-blown OAuth Authentication Service Provider ist out of scope for the wallabag server.
My suggestion would be:

Step 1:
Get the API auth right. (Native login -> create API credentials -> automatic get/renew short-lived token -> use API with token)

Step 2:
Add optional OAuth with an external auth service as alternative to native database credentials.

**Comment by deanishe** [comment]
> Why is that the front end must sport a native login mechanism?

Because OAuth is not an *authentication* mechanism. It's an *authorisation* mechanism. You need a separate authentication mechanism. Only after using that to identify yourself (logging in) can you use OAuth to authorise other systems to access your data.

> The user has a “company account” and he delegates access to parts of his account to different apps in the company. All those apps are OAuth2 clients.

They might be. But you still can't log into the authentication server using OAuth. You can only use OAuth to *authorise other systems* to access your account data after you've logging into the authentication server *using some other mechanism*.

So using OAuth as the *authentication* mechanism for Wallabag only makes sense if you're talking about using Google or Facebook or some other service to log into Wallabag. **Not** if you're talking about Wallabag being the authentication service. Because OAuth isn't an authentication service.

> To put it simply:

> - You go to your instance URL
> - It redirects you to the built-in AS which displays an authentication page
> - You log in with your username and password
> - This saves a session cookie in your browser

None of that is OAuth. That would be the native authentication mechanism.

> - You then decide to install Read It Later (GTK client app)
> - You launch Read It Later which prompts you for your instance URL
> - You type in your instance URL and validate
> - This launches your browser to Wallabag built in AS’ authentication page

You have this bit wrong. OAuth doesn't send you to the authentication page (the login). It sends you to the authorisation page (let app XYZ access your data). You might *also* have to visit the authentication (login) page if you aren't already logged in **but that is not part of the OAuth workflow.** OAuth has *nothing* to say about how you perform authentication.

> If the frontend keeps its own authentication interface independant of OAuth2, but we implement the built in AS for third parties client, then you can

That is exactly how OAuth works. You have your own authentication mechanism and then use OAuth to *authorise* 3rd-party apps.

> If the frontend is considered as an OAuth2 client too, you can:

> - do all of the above
> - plus adding new third party clients on the same machine doesn’t require you to log in again

This has nothing to do with OAuth. This is just sharing access tokens.

**Comment by thibaultamartin** [comment]
@deanishe I feel we should discuss this issue in private, since this is really getting into a purist argument, and I don't especially need to convince you except if you wish to contribute :)

I have checked with seasoned OAuth2 experts and architects I worked with. To their eyes, Dynamic Client registration, Auth Code + PKCE and considering the frontend as an OAuth2 clients is valid, OAuth2 compliant, and is conform with the best practices, so I'm going with that.

If we were to continue in private, I think you may want to double check the Authorization Code flow in details. There is no token sharing involved in the scenario described above.

**Comment by deanishe** [comment]
> this is really getting into a purist argument

It really isn't. The difference between authentication and authorisation is absolutely critical to understanding what OAuth is good for and what it isn't.

> I have checked with seasoned OAuth2 experts and architects I worked with […] is conform with the best practices

It is. But not for the purpose you seem to want to use it for. Please ask your expert colleagues to explain authentication vs authorisation and how OAuth fits in. I'm sure they'll tell you exactly what 
I've said: you still need a native login mechanism.

> I think you may want to double check the Authorization Code flow in details

Please note it's **Authorization**, not authentication. There is no mention of authentication in the whole flow because OAuth does not do authentication.

> If we were to continue in private

I don't see any point, tbh. I've tried to explain as clearly as I can that OAuth is not an authentication mechanism, so you'll still need to write your own, but we're apparently just talking past each other.

**Comment by 00dani** [comment]
It may be worth mentioning that OAuth's most common flow, the "authorization code" flow, is actually perfectly usable for command-line applications and the like. There's a standard redirect URI that these types of apps should use: `urn:ietf:wg:oauth:2.0:oob`, with "oob" meaning "out of band".

When this special URI is used, rather than actually redirecting anywhere, a supporting authorisation server generates a page displaying the authorisation code to the user directly. The code can then be copy-pasted into the command-line app, allowing the usual OAuth flow to proceed from there. It's equivalent in security to using a normal redirect URI, since the app must still provide its client ID and secret when consuming the new authorisation code. (In other words, pasting the code into the wrong place by accident doesn't actually give your access token to the wrong app.)

There is also the "loopback" approach: the command-line app simply starts up a tiny HTTP server on localhost:some_random_port, and then that server is used as the redirect URI. This allows the command-line app to grab the authorisation code directly from the redirected request without any user intervention, but it's more work on the CLI app dev's part and usually isn't much more convenient than a copy-paste job for the typical user of command-line apps.

If refresh tokens worked properly, either one of these OAuth approaches could be applied as a fast one-time configuration step when first setting up your command-line app - the app would simply save the access and refresh tokens in its configuration and use those later on, refreshing and replacing the access token when required.

Either way it's much better than giving your command-line app both its own client ID and client secret, *and* your username and password.

**Comment by mediaformat** [comment]
Here is a scenario, you use your Kindle with KOReader for un-distracted reading. 
Problem is the username and password are stored cleartext in a file.

This is where generating an API key linked to a limited scope would be ideal.

**Comment by ataraxus** [comment]
Long time lurker here:
exactly this! I have built a small service which grabs articles from my wallabag, converts them and send them to my kindle. Its quite alien to me to have my user credentials addtionally to my api secret in that configuration. 

Would be nice if it were possible to use only api secrets for automated services.

**Comment by virtadpt** [comment]
Same.  I had to build an entire agent workflow to go back and forth with Wallabag a couple of times to get a token.  Why not just have an API token that can be used in API calls like everybody else?

**Comment by setop** [comment]
this issue is the root of all client apps issues : https://github.com/wallabag/wallabagger/issues/145 and https://github.com/wallabag/wallabag/issues/2800

There must be a way :sob:

**Comment by mediaformat** [comment]
https://github.com/wallabag/wallabag/pull/3068 👀

**Comment by cheywood** [comment]
Seven years! Having an API without this gaping hole will do well for the health of the ecosystem and thus project once it's finally plugged. Please please please ❤️ :pray:

**Comment by Altair31** [comment]
To be honest, I stopped using wallabag for this unique reason

**Comment by janxkoci** [comment]
Wow, that was a crazy read 😂 

For start, I just looked up OAuth2 and the [first link](https://auth0.com/intro-to-iam/what-is-oauth-2) said in like third paragraph:

> OAuth 2.0 is an authorization protocol and NOT an authentication protocol.

So I hope that's clear now.

Then there were two guys having confused intermezzo because one thought "clients" were "other users" while in reality they were "apps used by the user" - hilarious exchange, just not sure why it was happening here.

So to sum up - we still need a proper login mechanism, and possibly oauth, maybe, who knows. 

I know nothing about PHP, but everyone is vibe-coding these days and even though I despise the whole concept, I despise having my passwords for grabs even more, so maybe one of us will have to give the vibes a shot! Unless a real dev steps up and gets it done, of course..

PS: I've joined in August as an ex-Pocket user, so at the moment I'm still willing to give wallabag a shot, but this issue being 7 years old doesn't make it look good. Let's not forget this entire thing is a non-issue with Instapaper (there the issue is paying for search, which I find absurd).

**Comment by janxkoci** [comment]
Also, one of you above mentioned this:

> Wallabag is not a critical app and shouldn't be the main target in case of threat.

From my point of you, my Wallabag account has more personal information about me than my Facebook account. All the interesting things I've found and saved over the years in my Pocket and last month imported to Wallabag - before learning about this issue - are a literal gold mine for marketing teams of the world, which makes our accounts potentially lucrative for hackers.

**Comment by zargex** [comment]
What appened with https://github.com/wallabag/wallabag/pull/3068 ? it was just closed and deleted.

I also think that the current oauth system doesn't fit the nature of wallabag (probably several self hosting app have similar issues). Asking for username and password in an oauth2 client means that something is wrong. 

It would be better to have just API token.

(Edit: Typo)
